File: python/src/agent_squad/neutrosophic/scorer.py
import math
import re

from agent_squad.neutrosophic.triplet import Triplet

HEDGING_PATTERNS = (
    "maybe",
    "might",
    "possibly",
    "probably",
    "not sure",
    "unclear",
    "unknown",
    "ambiguous",
    "insufficient",
    "could be",
    "it depends",
)

ERROR_PATTERNS = (
    "error",
    "failed",
    "failure",
    "cannot",
    "can't",
    "unable",
    "invalid",
    "contradiction",
    "conflict",
    "not possible",
    "i do not know",
    "i don't know",
)

REFUSAL_PATTERNS = (
    "i can't help",
    "i cannot help",
    "i cannot comply",
    "i can't comply",
    "sorry, but i can't",
    "sorry, but i cannot",
)


def score_text_response(text: str) -> Triplet:
    """Score plain text with deterministic baseline heuristics.

    This scorer is intentionally lightweight. It provides a dependency-free
    baseline until framework integrations can inject model-specific scorers.
    """
    if not isinstance(text, str):
        raise TypeError("text must be a string")

    normalized = " ".join(text.lower().split())
    if not normalized:
        return Triplet(T=0.0, I=1.0, F=0.0)

    hedge_count = _count_matches(normalized, HEDGING_PATTERNS)
    error_count = _count_matches(normalized, ERROR_PATTERNS)
    refusal_count = _count_matches(normalized, REFUSAL_PATTERNS)
    word_count = len(re.findall(r"\b\w+\b", normalized))

    substance = min(word_count / 80, 1.0)
    directness = 1.0 if hedge_count == 0 else max(0.0, 1.0 - (hedge_count * 0.2))

    truth = _clamp((0.35 + (0.45 * substance) + (0.20 * directness)) - (0.20 * error_count))
    indeterminacy = _clamp((0.15 if word_count >= 8 else 0.45) + (0.18 * hedge_count))
    falsity = _clamp((0.22 * error_count) + (0.35 * refusal_count))

    return Triplet(T=truth, I=indeterminacy, F=falsity)


def score_classifier_confidence(confidence: float, selected: bool) -> Triplet:
    """Map legacy classifier confidence into a neutrosophic triplet."""
    if isinstance(confidence, bool):
        raise TypeError("confidence must be a number")

    confidence_value = float(confidence)
    if not math.isfinite(confidence_value):
        raise ValueError("confidence must be finite")

    confidence_value = _clamp(confidence_value)
    if selected:
        return Triplet(T=confidence_value, I=1 - confidence_value, F=0)

    return Triplet(T=0, I=max(1 - confidence_value, 0.7), F=0)


def _count_matches(text: str, patterns: tuple[str, ...]) -> int:
    return sum(
        len(re.findall(rf"(?<!\w){re.escape(pattern)}(?!\w)", text))
        for pattern in patterns
    )


def _clamp(value: float) -> float:
    return max(0.0, min(1.0, value))

File: typescript/src/neutrosophic/scorer.ts
import { Triplet } from "./triplet";

const HEDGING_PATTERNS = [
  "maybe",
  "might",
  "possibly",
  "probably",
  "not sure",
  "unclear",
  "unknown",
  "ambiguous",
  "insufficient",
  "could be",
  "it depends",
];

const ERROR_PATTERNS = [
  "error",
  "failed",
  "failure",
  "cannot",
  "can't",
  "unable",
  "invalid",
  "contradiction",
  "conflict",
  "not possible",
  "i do not know",
  "i don't know",
];

const REFUSAL_PATTERNS = [
  "i can't help",
  "i cannot help",
  "i cannot comply",
  "i can't comply",
  "sorry, but i can't",
  "sorry, but i cannot",
];

export function scoreTextResponse(text: string): Triplet {
  if (typeof text !== "string") {
    throw new TypeError("text must be a string");
  }

  const normalized = text.toLowerCase().split(/\s+/).filter(Boolean).join(" ");
  if (!normalized) {
    return new Triplet(0.0, 1.0, 0.0);
  }

  const hedgeCount = countMatches(normalized, HEDGING_PATTERNS);
  const errorCount = countMatches(normalized, ERROR_PATTERNS);
  const refusalCount = countMatches(normalized, REFUSAL_PATTERNS);
  const wordCount = normalized.match(/\b\w+\b/g)?.length || 0;
  const substance = Math.min(wordCount / 80, 1.0);
  const directness = hedgeCount === 0 ? 1.0 : Math.max(0.0, 1.0 - hedgeCount * 0.2);

  const truth = clamp(0.35 + 0.45 * substance + 0.20 * directness - 0.20 * errorCount);
  const indeterminacy = clamp((wordCount >= 8 ? 0.15 : 0.45) + 0.18 * hedgeCount);
  const falsity = clamp(0.22 * errorCount + 0.35 * refusalCount);

  return new Triplet(truth, indeterminacy, falsity);
}

export function scoreClassifierConfidence(confidence: number, selected: boolean): Triplet {
  if (typeof confidence !== "number") {
    throw new TypeError("confidence must be a number");
  }
  if (!Number.isFinite(confidence)) {
    throw new RangeError("confidence must be finite");
  }

  const confidenceValue = clamp(confidence);
  if (selected) {
    return new Triplet(confidenceValue, 1 - confidenceValue, 0);
  }

  return new Triplet(0, Math.max(1 - confidenceValue, 0.7), 0);
}

function countMatches(text: string, patterns: string[]): number {
  return patterns.reduce((count, pattern) => {
    const escaped = pattern.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    return count + (text.match(new RegExp(`(?<!\\w)${escaped}(?!\\w)`, "g"))?.length || 0);
  }, 0);
}

function clamp(value: number): number {
  return Math.max(0.0, Math.min(1.0, value));
}
